File: browser/src/js/moneysocket/core/message/notification/preimage.js
// Copyright (c) 2020 Jarret Dyrbye
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php


const BinUtl = require('../../../utl/bin.js').BinUtl;
const MoneysocketNotification = require(
    './notification.js').MoneysocketNotification;
let NOTIFICATION_SUBCLASSES = require(
    './notification.js').NOTIFICATION_SUBCLASSES;

class NotifyPreimage extends MoneysocketNotification {
    constructor(preimage, ext, request_reference_uuid) {
        super("NOTIFY_PREIMAGE", request_reference_uuid);
        this.preimage = preimage;
        this.ext = ext
    }

    cryptLevel() {
        return "AES";
    }

    static castClass(msg_dict) {
        var c = new NotifyPreimage(msg_dict['bolt11'],
                                   msg_dict['ext'],
                                   msg_dict['request_reference_id']);
        Object.keys(msg_dict).forEach(key => {
            c[key] = msg_dict[key];
        });
        return c;
    }

    static checkValidMsgDict(msg_dict) {
        if (! ('preimage' in msg_dict)) {
            return "no preimage included";
        }
        if (typeof msg_dict['preimage'] != 'string') {
            return "unknown preimage type";
        }
        if (! BinUtl.stringIsHex(msg_dict['preimage'])) {
            return "preimage not hex string";
        }
        if (msg_dict['preimage'].length != 64) {
            return "preimage not 256-bit value hex string";
        }
        if (msg_dict['ext'] != null) {
            if (typeof msg_dict['ext'] != 'string') {
                return "unknown ext field type";
            }
            if (msg_dict['ext'].length > 4096) {
                return "ext is longer than 4096 characters";
            }
        }
        return null;
    }
}

NOTIFICATION_SUBCLASSES['NOTIFY_PREIMAGE'] = NotifyPreimage;

exports.NotifyPreimage = NotifyPreimage;

File: python/moneysocket/core/message/notification/preimage.py
# Copyright (c) 2020 Jarret Dyrbye
# Distributed under the MIT software license, see the accompanying
# file LICENSE or http://www.opensource.org/licenses/mit-license.php

import string
import json

from moneysocket.core.message.notification.notification import (
    MoneysocketNotification)

class NotifyPreimage(MoneysocketNotification):
    MUST_BE_CLEARTEXT = False

    def __init__(self, preimage, ext=None, request_reference_uuid=None):
        super().__init__("NOTIFY_PREIMAGE",
                         request_reference_uuid=request_reference_uuid)
        self['preimage'] = preimage
        self['ext'] = ext

    @staticmethod
    def cast_class(msg_dict):
        c = NotifyPreimage(msg_dict['preimage'],
                           msg_dict['ext'],
                           msg_dict['request_reference_uuid'])
        c.update(msg_dict)
        return c

    @staticmethod
    def check_valid_msg_dict(msg_dict):
        if 'preimage' not in msg_dict.keys():
            return "no preimage included"
        if type(msg_dict['preimage']) != str:
            return "unknown preimage type"
        if not all(c in string.hexdigits for c in msg_dict['preimage']):
            return "preimage not hex string"
        if len(msg_dict['preimage']) != 64:
            return "preimage not 256-bit value hex string"
        if msg_dict['ext'] != None:
            if type(msg_dict['ext']) != str:
                return "unknown ext type"
            if len(msg_dict['ext']) > 4096:
                return "ext is larger than 4096 characters"
        return None


MoneysocketNotification.NOTIFICATION_SUBCLASSES['NOTIFY_PREIMAGE'] = (
    NotifyPreimage)

